### PR TITLE
Fix diagonal

### PR DESCRIPTION
**Pull request by fukatani**
#427

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -3026,6 +3026,8 @@ cpdef _scatter_op(ndarray a, slices, value, op):
 
 cpdef ndarray _diagonal(ndarray a, Py_ssize_t offset=0, Py_ssize_t axis1=0,
                         Py_ssize_t axis2=1):
+    axis1 = axis1 % a.ndim
```
Comment (okuta) [inline cupy/core/core.pyx]: Do we need to check out of range?

**Comment by fukatani** [comment]
Thanks! I added range check and its test.

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -3026,6 +3026,13 @@ cpdef _scatter_op(ndarray a, slices, value, op):
 
 cpdef ndarray _diagonal(ndarray a, Py_ssize_t offset=0, Py_ssize_t axis1=0,
                         Py_ssize_t axis2=1):
+    if (axis1 >= a.ndim or abs(axis2) >= a.ndim or
```
Comment (okuta) [inline cupy/core/core.pyx]: Please assign `a.ndim` to local variable for performance up.

**Comment by fukatani** [comment]
OK! I addressed your comment.

**Comment by okuta** [comment]
Thanks.
Jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**